mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java

```diff
@@ -21,76 +18,122 @@
+        return new ConnectionContext<Cluster>() {
+
+            Cluster cluster;
+
+            @Override
+            public Cluster getStorage() {
+                if (cluster == null) {
+                    synchronized (this) {
+                        if (cluster == null) {
+                            cluster = CouchbaseCluster.create(builder, contactPoint);
+                        }
+                    }
+                }
+                return cluster;
+            }
+
+            @Override
+            public void close() throws Exception {
+                if (cluster != null) {
+                    synchronized (this) {
+                        if (cluster != null) {
+                            cluster.disconnect();
+                            cluster = null;
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * @return A builder for a {@link CouchbaseProvisioner}.
+     */
+    public static ClusterBuilder couchbase() {
+        return new CouchbaseProvisionerBuilder();
+    }
+
     @Override
     public <K, V> Mache<K, V> getCache(Class<K> keyType, Class<V> valueType) {
         return new MacheFactory().create(getCacheLoader(keyType, valueType));
     }
-    
+
     @Override
     public <K, V> AbstractCacheLoader<K, V, ?> getCacheLoader(Class<K> keyType, Class<V> valueType) {
-    	return new CouchbaseCacheLoader<>(keyType, valueType, bucketSettings,
-                couchbaseEnvironment, nodes, adminUser, adminPassword, schemaOptions);
+        return new CouchbaseCacheLoader<>(keyType, valueType, bucketSettings, connectionContext, adminUser, adminPassword, schemaOptions);
     }
 
     /**
-     * @return A builder for a {@link CouchbaseProvisioner}.
+     * Forces bucket settings to be provided.
      */
-    public static BucketBuilder couchbase() {
-        return CouchbaseProvisionerBuilder::new;
+    public interface ClusterBuilder {
+        BucketBuilder withContext(ConnectionContext<Cluster> connectionContext);
     }
 
-    /**
-     * Forces bucket settings to be provided.
-     */
     public interface BucketBuilder {
-        CouchbaseProvisionerBuilder withBucketSettings(BucketSettings bucketSettings);
+        AdminSettings withBucketSettings(BucketSettings bucketSettings);
+    }
+
+    public interface AdminSettings {
+        SchemaOptionsSettings withAdminDetails(String adminUser, String adminPassword);
+
+        SchemaOptionsSettings withDefaultAdminDetails();
+    }
+
+    public interface SchemaOptionsSettings {
+        CouchbaseProvisionerBuilder withSchemaOptions(SchemaOptions schemaOptions);
+
+        CouchbaseProvisionerBuilder withDefaultSchemaOptions();
     }
 
     /**
      * A builder with defaults for a Couchbase cluster.
      */
-    public static class CouchbaseProvisionerBuilder {
+    public static class CouchbaseProvisionerBuilder implements ClusterBuilder, BucketBuilder, AdminSettings, SchemaOptionsSettings {
```

Comment:
We've kinda broken what was intended here. We use the lamba builder pattern to enforce the user to provide mandatory fields, which are then provided to this builder. This builder has a set of defaults that the user may change, and when they're finished just call build.